kubetest/federation.go

```diff
@@ -27,3 +56,294 @@ func FedUp() error {
+
+	return l.readModifyWrite(func(content []byte) ([]byte, bool) {
+		if bytes.HasSuffix(content, barrier) {
+			return nil, true
+		}
+		return append(content, []byte(line)...), false
+	})
+}
+
+func (l *gcsLock) unlock(data string) error {
+	log.Printf("Releasing lock: %s", data)
+
+	return l.readModifyWrite(func(content []byte) ([]byte, bool) {
+		return removeLineWithPrefix(content, []byte(data)), false
+	})
+}
+
+func (l *gcsLock) addBarrier() error {
+	log.Printf("Adding barrier")
+
+	return l.readModifyWrite(func(content []byte) ([]byte, bool) {
+		return append(content, []byte(barrier)...), false
+	})
+}
+
+func (l *gcsLock) waitForEmptyBarrier() error {
+	log.Printf("Waiting for empty barrier")
+
+	obj := l.client.Bucket(l.bucket).Object(l.obj)
+
+	return poll(func() (bool, error) {
+		content, err := readObject(l.ctx, obj)
+		if err != nil {
+			return false, fmt.Errorf("failed to read object %s/%s: %v", l.bucket, l.obj, err)
+		}
+		if bytes.Equal(content, barrier) {
+			return true, nil
+		}
+		return false, nil
+	})
+}
+
+func (l *gcsLock) removeBarrier() error {
+	log.Printf("Removing barrier")
+
+	return l.readModifyWrite(func(content []byte) ([]byte, bool) {
+		return removeLineWithPrefix(content, barrier), false
+	})
+}
+
+func (l *gcsLock) readModifyWrite(modify func([]byte) ([]byte, bool)) error {
+	obj := l.client.Bucket(l.bucket).Object(l.obj)
+
+	return poll(func() (bool, error) {
+		log.Print("Read-Modify-Write")
+
+		attrs, err := obj.Attrs(l.ctx)
+		if err != nil && err != storage.ErrObjectNotExist {
+			return false, fmt.Errorf("failed to retrieve attributes for %s/%s: (%T)%v", l.bucket, l.obj, err, err)
+		}
+
+		content := make([]byte, 0)
+		conditions := storage.Conditions{}
+		if err == storage.ErrObjectNotExist {
+			conditions.DoesNotExist = true
+			log.Printf("object %s/%s doesn't exist, will be created in the next write cycle", l.bucket, l.obj)
+		} else {
+			conditions.GenerationMatch = attrs.Generation
+
+			log.Printf("reading object %s/%s", l.bucket, l.obj)
+			content, err = readObject(l.ctx, obj)
+			if err != nil {
+				return false, fmt.Errorf("failed to read object %s/%s: %v", l.bucket, l.obj, err)
+			}
+		}
+
+		newContent, retry := modify(content)
+		if retry {
+			return false, nil
+		}
+
+		w := obj.If(conditions).NewWriter(l.ctx)
+		if _, err := w.Write(newContent); err != nil {
+			return false, fmt.Errorf("failed to write the object %s/%s: %v", l.bucket, l.obj, err)
+		}
+
+		err = w.Close()
+		if apiErr, ok := err.(*googleapi.Error); ok {
+			if apiErr.Code == http.StatusPreconditionFailed && apiErr.Errors[0].Reason == conditionNotMet {
+				return false, nil
+			} else {
+				return false, fmt.Errorf("failed to close the writer for %s/%s: %v", l.bucket, l.obj, apiErr)
+			}
+		} else if err != nil {
+			return false, fmt.Errorf("failed to close the writer for %s/%s: %v", l.bucket, l.obj, err)
+		}
+		return true, nil
+	})
+}
+
+func poll(cond func() (bool, error)) error {
+	ok, err := cond()
```

Comment:
Does NewTicker not return a ticker that fires an initial burst? That's annoying.